File: 2021-09/2021-9-29 517.超级洗衣机/FindMinMoves.cs
using System;
using System.Linq;

public class Solution
{
    public int FindMinMoves(int[] machines)
    {
        // 求和 找到平均数 
        // 如果不能平均就 return -1
        // 如果一个机子已经达到平均，就传送到其他洗衣机
        var sum = machines.Sum();
        if (sum % machines.Length != 0)
        {
            return -1;
        }

        var mean = sum / machines.Length;

        var result = 0;
        var count = 0;
        foreach (var machine in machines)
        {
            var temp = machine - mean;
            // 因为每次只能移动相邻的，所以要叠加
            count += temp;
            // 需要移入的和需要移除的里面找一个最大值
            result = Math.Max(result, Math.Max(Math.Abs(count), temp));
        }

        return result;
    }
}

File: 2021/2021-09/2021-9-29 517.超级洗衣机/findMinMoves.go
package main

import (
	"math"
)

func main() {
	machines := []int{1, 0, 5}
	findMinMoves(machines)
}

func findMinMoves(machines []int) int {
	var temp, res int

	length, sum := machinesSum(machines)
	if sum%length != 0 {
		return -1
	}

	avg := sum / length
	for _, value := range machines {
		temp += value - avg
		// 从左右的洗衣机中取衣服和当前洗衣机需要转移的衣服取最大值
		res = max(res, max(int(math.Abs(float64(temp))), int(math.Abs(float64(value-avg)))))
	}
	return res
}

func machinesSum(machines []int) (int, int) {
	length, sum := 0, 0
	for _, value := range machines {
		length++
		sum += value
	}
	return length, sum
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}
